internal/cmd/oohelperd/internal/nwcth/measure.go

```diff
@@ -0,0 +1,281 @@
+	endpoint string,
+	wg *sync.WaitGroup,
+	result *MeasureEndpointResult,
+) {
+	URL, _ := url.Parse(creq.HTTPRequest)
+	httpMeasurement := CtrlHTTPMeasurement{Endpoint: endpoint, Protocol: URL.Scheme}
+	var conn net.Conn
+	conn, httpMeasurement.TCPConnect = TCPDo(ctx, &TCPConfig{
+		Endpoint: endpoint,
+	})
+	if conn == nil {
+		return
+	}
+	defer conn.Close()
+	var transport http.RoundTripper
+	switch URL.Scheme {
+	case "http":
+		transport = nwebconnectivity.GetSingleTransport(conn, nil)
+	case "https":
+		var tlsconn *tls.Conn
+		cfg := &tls.Config{ServerName: URL.Hostname()}
+		tlsconn, httpMeasurement.TLSHandshake = TLSDo(ctx, &TLSConfig{
+			Conn:     conn,
+			Endpoint: endpoint,
+			Cfg:      cfg,
+		})
+		if tlsconn == nil {
+			return
+		}
+		transport = nwebconnectivity.GetSingleTransport(tlsconn, cfg)
+	}
+	// perform the HTTP request: this provides us with the HTTP request result and info about HTTP redirection
+	httpMeasurement.HTTPRequest, result.httpRedirect = HTTPDo(ctx, &HTTPConfig{
+		Jar:       creq.HTTPCookieJar,
+		Headers:   creq.HTTPRequestHeaders,
+		Transport: transport,
+		URL:       URL,
+	})
+	// find out of the host also supports h3 support, which is announced in the Alt-Svc Header
+	h3Support := discoverH3Server(httpMeasurement.HTTPRequest, URL)
+	if h3Support != "" {
+		quicURL, _ := url.Parse(URL.String())
+		quicURL.Scheme = h3Support
+		result.h3Location = quicURL.String()
+	}
+	result.CtrlEndpoint = &httpMeasurement
+}
+
+func measureH3(
+	ctx context.Context,
+	creq *CtrlRequest,
+	endpoint string,
+	wg *sync.WaitGroup,
+	result *MeasureEndpointResult,
+) {
+	URL, _ := url.Parse(creq.HTTPRequest)
+	h3Measurement := CtrlH3Measurement{Endpoint: endpoint, Protocol: URL.Scheme}
+	var sess quic.EarlySession
+	tlscfg := &tls.Config{ServerName: URL.Hostname(), NextProtos: []string{URL.Scheme}}
+	qcfg := &quic.Config{}
+	sess, h3Measurement.QUICHandshake = QUICDo(ctx, &QUICConfig{
+		Endpoint:  endpoint,
+		QConfig:   qcfg,
+		TLSConfig: tlscfg,
+	})
+	if sess == nil {
+		return
+	}
+	transport := nwebconnectivity.GetSingleH3Transport(sess, tlscfg, qcfg)
+	h3Measurement.HTTPRequest, result.httpRedirect = HTTPDo(ctx, &HTTPConfig{
+		Jar:       creq.HTTPCookieJar,
+		Headers:   creq.HTTPRequestHeaders,
+		Transport: transport,
+		URL:       URL,
+	})
+	sess.CloseWithError(0, "")
+	result.CtrlEndpoint = &h3Measurement
+}
+
+func mergeEndpoints(addrs []string, clientAddrs []string) (out []string) {
+	unique := make(map[string]bool, len(addrs)+len(clientAddrs))
+	for _, a := range addrs {
+		unique[a] = true
+	}
+	for _, a := range clientAddrs {
+		unique[a] = true
+	}
+	for key := range unique {
+		out = append(out, key)
+	}
+	return out
+}
+
+// getEndpoints connects IP addresses with the port associated with the URL scheme
+func getEndpoints(addrs []string, URL *url.URL) []string {
+	out := []string{}
+	if URL.Scheme != "http" && URL.Scheme != "https" && URL.Scheme != "h3" && URL.Scheme != "h3-29" {
```

Comment:
I think we need an extensible way to add/remove supported QUIC versions: we have a bunch of places where we check for the scheme and it may be more convenient to have a single place where to change that.

Reply:
I introduced a map to manage the supported QUIC versions. It lives in `measure.go`